Allow querying of stack state snapshots

hausdorff opened the pull request.

chrsmith commented on pkg/backend/backend.go:
```diff
@@ -222,3 +222,30 @@ func (c *backendClient) GetStackOutputs(ctx context.Context, name string) (resou
+		return nil, err
+	}
+	if s == nil {
+		return nil, errors.Errorf("unknown stack \"%s\"", name)
```
Nit: You can use `%q` as a format specifier, which will accept a string and print it with quotation marks around it. (So you don't have to add the `\"` yourself.

chrsmith commented on pkg/resource/deploy/builtins.go:
```diff
@@ -127,22 +127,33 @@ func (p *builtinProvider) Delete(urn resource.URN, id resource.ID,
 
 func (p *builtinProvider) Invoke(tok tokens.ModuleMember,
 	args resource.PropertyMap) (resource.PropertyMap, []plugin.CheckFailure, error) {
+	if tok != "pulumi:pulumi:readStackResourceOutputs" {
```
Should `"Pulumi:pulumi:readStackResourceOutputs"` be defined as a `const` somewhere else in the package? Like with all the other definitions of well-known `tokens.ModuleMember`s?

chrsmith commented on pkg/resource/deploy/plan.go:
```diff
@@ -35,6 +35,10 @@ import (
+
+	// GetStackResourceOutputs returns the resource outputs (if any) for the named stack or an
+	// error if the stack cannot be found.
+	GetStackResourceOutputs(ctx context.Context, stackName, typ string) (resource.PropertyMap, error)
```
It isn't immediately obvious what the `typ` parameter does. From the looks of it, it is an optional filter, and if non-empty `GetStackResourceOutputs` will only return the stack outputs with that type? If so, we should probably mention it -- even if as a pithy "takes an optional type filter" -- in the comments.

hausdorff commented:
@chrsmith Feedback addressed.

A couple other things that are worth explicitly discussing here.

* **Should we expose the entire state snapshot through `Invoke`?** In this PR we choose to expose only the _resource outputs_ through `Invoke` -- that is, the live objects as they exist in the last state snapshot. I wasn't sure of the implications so I was conservative, but either are pretty easy to do.
* **Should we expose this `Invoke` call through both the Python and TS SDKs?** In this PR we don't do either -- we meant to expose this in an upcoming PR. This is also fairly easy to do, but we'd have to implement an `EnumerablePromise` per the design doc.

Answers to these, I think, would be enough to move us out of the "draft PR" stage.

hausdorff commented:
oh cc @lukehoban @joeduffy

joeduffy commented:
> Should we expose this Invoke call through both the Python and TS SDKs?

We should. For core features, we're trying to get out of the habit of letting the SDKs get out of synch, since that leads to a "downstream" need to catch up, often after a user trips over the delta.

lukehoban commented:
> I'd expose only the minimal necessary for now, until we have a scenario motivating more. Just my $0.02. 

I agree.  In particular, the state file itself is a carefully versioned artifact, which may not have 100% backward compatibility over time (we take on responsibility in the engine and service to migrate between versions).  I don't think we should push that complexity onto users.  The Pulumi property bag model of resources is something that we can likely expose more directly without risk of breaking changes.  More generally - we likely want the interface here to version with the RPC model, not with the engine/checkpoint file version.

lukehoban commented:
> Should we expose this Invoke call through both the Python and TS SDKs?

I don't exactly follow this one.  I assume that either in this PR or a follow-up we'll need to expose this via the query APIs?

hausdorff commented:
@lukehoban @joeduffy ok to recap:

* Let's expose this through both the Python and TS/JS SDKs.
* Let's have the `Invoke` return only the resource outputs for now, rather than the whole state checkpoint.

I think this is ready for review, then?

ellismg commented on pkg/backend/backend.go:
```diff
@@ -230,3 +230,33 @@ func (c *backendClient) GetStackOutputs(ctx context.Context, name string) (resou
+	pm := resource.PropertyMap{}
+	for _, r := range snap.Resources {
+		if typ == "" || string(r.Type) == typ {
+			pm[resource.PropertyKey(r.ID)] = resource.NewObjectProperty(r.Outputs)
```
Is it guaranteed that `r.ID` is unique across all resources in a snapshot? I would not expect that to be the case in general, but maybe part of our system enforces it?

lukehoban replied:
No - `r.ID` is not unique.  `URN`s are almost unique, but may need to skip any pending delete entries to ensure uniqueness.

ellismg commented:
Regarding this:

> Please also check with @ellismg that we're being careful about secrets in the necessary ways.

We will want to ensure that when you get a property back from the map that we project into the language, the type of that property is an `Output<whatever>`.  The secrets work will introduce a new type of `PropertyValue` that wraps another PropertyValue and the SDKs will understand how to deserialize this into an `Output<T>`.  At the level of the code you're touching here, there shouldn't be anything you need to do for secrets.

lukehoban commented on pkg/backend/httpstate/backend.go:
```diff
@@ -1255,3 +1255,33 @@ func (c httpstateBackendClient) GetStackOutputs(ctx context.Context, name string
+		if typ == "" || string(r.Type) == typ {
+			pm[resource.PropertyKey(r.ID)] = resource.NewObjectProperty(r.Outputs)
+		}
+	}
```
Do we need two implementations of this?  Can't we just forward to the underlying backend like we do for `GetStackOutputs`?

hausdorff replied:
Totally not necessary -- very WIP. We will fix.

hausdorff commented:
@ellismg @lukehoban I believe we've addressed your feedback about the new `Invoke` call.

I've merged this with #2649, pulling in the `EnumerablePromise` implementation. With this implementation, it is possible to write code like this:

```typescript
pulumi.runtime
    .listResourceOutputs(undefined, aws.rds.Instance.isInstance)
    .filter(rds => parseInt(rds.engineVersion) < 5.7);

pulumi.runtime
    .listResourceOutputs(undefined, isRdsInstance)
    .filter(group =>
        group.ingress.find(rule => rule.cidrBlocks.find((block: any) => block == "0.0.0.0/0")),
    );
```

And so on. I'm still in the process of testing the `Queryable` type stuff -- after that, I'll start to implement the `pulumi query` command.

ellismg commented:
> @ellismg @lukehoban I believe we've addressed your feedback about the new Invoke call.

Did you push?  I'm not seeing any changes here and #2649 appears to be closed now.

hausdorff commented:
@ellismg oh, lol, you're right. I pushed to a different branch. :)

justinvp commented on sdk/nodejs/query/interfaces.ts:
```diff
@@ -0,0 +1,44 @@
+// model.
+//
+
+export interface Enumerable<T> {
```
Instead of defining our own `Enumerable`/`Enumerator` interfaces, why not build on ECMAScript 2015's `Iterable`/`Iterator` protocols?

 - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols
 - https://github.com/Microsoft/TypeScript/blob/342616781e80202f40872f044605806ba73837a3/src/lib/es2015.iterable.d.ts

```typescript
interface IteratorResult<T> {
    done: boolean;
    value: T;
}

interface Iterator<T> {
    next(value?: any): IteratorResult<T>;
    return?(value?: any): IteratorResult<T>;
    throw?(e?: any): IteratorResult<T>;
}

interface Iterable<T> {
    [Symbol.iterator](): Iterator<T>;
}
```

Then we could define our own `QueryableIterable<T>` and `QueryableIterablePromise<T>` interfaces that extend these to hang the query functions on.

hausdorff replied:
@justinvp I'm open to implementing it using iterators, but the truth is that when I read the docs I came away with no confidence that I understood how they're implemented, or what the technical implications of betting this on them would be. Since `Enumerable` is such a small part of the implementation here, I decided to just write it myself.

Some example questions I'm still confused about are: Does the error channel behave the same way as LINQ's? The completion channel? How hard is tree-shaking with iterators? Is there any advantage to using these protocols other than just being able to use `for`?

justinvp replied:
@hausdorff These are good questions, that we should try to answer. My high-level feedback is that we should consider building on top of what the language provides, so that we can be as idiomatic as possible in a given language and fit-in well with its ecosystem. (And even potentially aligning with future language proposals, much in the same way that TypeScript does).

> Is there any advantage to using these protocols other than just being able to use `for`?

In addition to `for-of`, I believe it also enables destructuring and the spread operator (...) for iterables. There are proposals that expand on iterators that appear to very much align with our needs:

 - [Asynchronous Iterators for JavaScript](https://github.com/tc39/proposal-async-iteration)
 - [Iterator Helpers](https://github.com/tc39/proposal-iterator-helpers)

Implementing using iterators and/or async iterators would enable all sorts of third party "iterator helpers" to _just work_ with what we provide, without us having to provide an implementation or adapters ourselves.

hausdorff replied:
@justinvp do you think it's necessary to build the entire abstraction on top of iterators, or is it enough for our purposes to expose `toIterator()` and `fromIterator()` functions?

justinvp replied:
@hausdorff, I think we should consider building the entire abstraction on top of iterators, unless there are good reasons not to beyond initial confusion/questions about how to implement them. If there are good reasons not to, I'd love to hear them, in which case, holding off and potentially providing `toIterator()` and `fromIterator()` adapters, if there's a need, sounds reasonable. Is there anything I can do to help clarify the initial uncertainty?

Also, I assume we eventually want to expose this in Python, right? Would we build on top of Python iterators so that things like `itertools` would _just work_ with our APIs, or are we going to build-out our own query operators on top of our own abstraction? (Apologies if this has already been discussed in a design meeting that I missed).

hausdorff replied:
@justinvp ok, we've started the move to an iterator-based combinator approach, where every query operator (_e.g._, `map`, `filter`) is "just" an async iterator.

hausdorff commented:
Ah, oops -- I just pushed my feature branch to this PR. My bad. I'll pull those commits back out. @lukehoban @ellismg what else needs to be done to agree on this approach?

justinvp commented on sdk/nodejs/query/interfaces.ts:
```diff
@@ -0,0 +1,44 @@
+// model.
+//
+
+export interface Enumerable<T> {
```
If we do end up sticking with these interfaces... If I have an an `Enumerable<T>` object, how do I get its `Enumerator<T>` if I wanted to iterate over it manually with an enumerator? I see for `EnumerablePromise<T>` I can presumably get it from the resolved promise.

hausdorff replied:
We had not planned to support this. But, with the move to async iterators, we get it incidentally.

justinvp commented on sdk/nodejs/query/interfaces.ts:
```diff
@@ -0,0 +1,44 @@
+    map<U>(f: (t: T) => U): Enumerable<U>;
+    take(n: number): Enumerable<T>;
+
+    toArray(): Promise<T[]>;
```
If we do end up sticking with these interfaces... Is it by design that this returns `Promise<T[]>` instead of `T[]`? None of the other functions on this interface return promises.

hausdorff replied:
This is on purpose. Once you start to do stuff async, you can't "get out of" the `Promise`. The way you would use this is:

```
const myarr = foo.filter(...).toArray();
```

hausdorff commented:
I've started experimenting with basing the query work on async iterators. What I'm hoping to do is to have a very combinator-oriented approach, where it is very easy for people to use iterators and "lift" them into some `Operator` that can be passed into the query API. Something like this:

```typescript
export function myFilter<TSource>(f: (t: TSource, i?: number) => boolean) {
    return async function*(source: AsyncIterableIterator<TSource>) {
        for await (const [t, i] of zip(source, range(0))) {
            if (f(t, i)) {
                yield t;
            }
        }
    };
}

xs.map(...).flatMap(...).pipe(myFilter)
```

hausdorff commented:
@justinvp yes, sorry, just to be clear: I am not expecting you to figure out build errors. :) I meant this to be kind of WIP.

ellismg commented on sdk/nodejs/resource.ts:
```diff
@@ -269,6 +269,13 @@ export abstract class CustomResource extends Resource {
+     * classes that inherit from `CustomResource`.
+     */
+    // tslint:disable-next-line:variable-name
+    /* @internal */ public readonly __pulumiType: string;
```
Cyrus recently did a pass here where he made sure that `@internal` was inside a doccomment style comment (e.g. `/** `) because otherwise I think these members were not being elided from our docs.  Can you move the `@internal` decorator to the top of comment?

ellismg commented on sdk/nodejs/query/operators.ts:
```diff
@@ -0,0 +1,746 @@
+// Ordering operators.
+//
+
+// public static OrderedSequence<TSource> OrderBy<TSource, TKey>(
```
Can we remove all of these comments?  I understand that they are here because we are mimicking the LINQ surface area, but I find them very confusing in the context of typescript files.

ellismg commented on sdk/nodejs/query/queryable.ts:
```diff
@@ -0,0 +1,54 @@
+import { OutputInstance } from "../index";
+import { CustomResource } from "../resource";
+
+export type Queryable<T> = T extends Promise<infer U1>
```
It might be nice to have a comment to the effect of what all this typescript dark magic is intended to do.

ellismg commented on sdk/nodejs/runtime/resource.ts:
```diff
@@ -501,6 +503,38 @@ export function registerResourceOutputs(res: Resource, outputs: Inputs | Promise
+ * guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards)
+ * that specifies which resource types to list outputs of.
+ *
+ * @example pulumi.runtime.listResourceOutput(aws.s3.Bucket.isInstance)
```
this is actually an incorrect example, right?  It should be `pulumi.runtime.listResourceOutput(undefined, aws.s3.Bucket.isInstance)`?

ellismg commented on sdk/nodejs/runtime/resource.ts:
```diff
@@ -501,6 +503,38 @@ export function registerResourceOutputs(res: Resource, outputs: Inputs | Promise
+ *
+ * @example pulumi.runtime.listResourceOutput(aws.s3.Bucket.isInstance)
+ */
+export function listResourceOutputs<U extends CustomResource = any>(
```
I do wonder about the shape of this API.  The optional first parameter is kinda wonky, if we expect the common case to be running queries against the "current stack", which I feel like will be the common case?  I expect that in general, I'm going to have my Pulumi project and then one or more `.ts` files that have operational queries I'm interested in.  In that world, I'm going to run something like:

```
$ pulumi query findPublicBuckets.ts
```

or something like:

```
$ pulumi query -s production findPublicBuckets.ts
```

and the queries I am going to write are just going to elide the stackName, right?

I wonder if we should swap the order of these arguments?  Or add an arguments bag and move the stackName parameter into it?

hausdorff commented on pkg/resource/deploy/plan.go:
```diff
@@ -41,6 +40,11 @@ type BackendClient interface {
 	DownloadPlugin(ctx context.Context, plug workspace.PluginInfo) (io.ReadCloser, error)
+
+	// GetStackResourceOutputs returns the resource outputs (if any) for a stack, or an
+	// error if the stack cannot be found. Resources are retrieved from the latest stack snapshot,
```
I prefer to keep it consistent. :) LMK if you feel strongly to the contrary.

hausdorff commented:
@ellismg I think this is ~ready now. I've removed the query code and sourced that to `@pulumi/query`. I've also renamed `QueryableResource` -> `ResolvedResource` as I think it makes a bit more sense as a name -- see comment. The one piece of work we do need to do is to publish the `@pulumi/query` package with an actual version, so that this can take a dependency on it. I've left the `CHANGELOG` changes out since this is all behind a debug flag.

hausdorff commented on sdk/nodejs/queryable/index.ts:
```diff
@@ -0,0 +1,57 @@
+ * is useful primarily when we're querying over resource outputs (e.g., using
+ * `pulumi.runtime.listResourceOutputs`), and we expect all values to be present and fully-resovled.
+ */
+export type ResovledResource<T extends Resource> = Omit<Resolved<T>, "urn" | "getProvider">;
```
No reason, other than that query primitives against live non-pulumi-managed resources will not have it. Also if we really need it we can always put it in.

ellismg replied:
> No reason, other than that query primitives against live non-pulumi-managed resources will not have it

Of course!  Makes sense.

hausdorff commented on tests/integration/query/query_test.go:
```diff
@@ -0,0 +1,34 @@
+		Dir:          "step1",
+		Dependencies: []string{"@pulumi/pulumi"},
+		// Quick:        true,
+		EditDirs: []integration.EditDir{
```
We want to test that it errors out when we try to create resources in query mode.

hausdorff replied:
I added comments.

hausdorff commented on tests/integration/query/query_test.go:
```diff
@@ -0,0 +1,34 @@
+				Dir:           "step2",
+				Additive:      true,
+				QueryMode:     true,
+				ExpectFailure: true,
```
See above

praneetloke commented on sdk/nodejs/queryable/index.ts:
```diff
@@ -0,0 +1,57 @@
+export type ResovledResource<T extends Resource> = Omit<Resolved<T>, "urn" | "getProvider">;
+
+type Resolved<T> = T extends Promise<infer U1>
+    ? ResovledSimple<U1>
```
Shouldn't this be `Reso_lv_edSimple` and not `Reso_vl_edSimple`? Same with the next two lines as well.

ellismg replied:
:+1:

praneetloke commented on sdk/nodejs/queryable/index.ts:
```diff
@@ -0,0 +1,57 @@
+ * is useful primarily when we're querying over resource outputs (e.g., using
+ * `pulumi.runtime.listResourceOutputs`), and we expect all values to be present and fully-resovled.
+ */
+export type ResovledResource<T extends Resource> = Omit<Resolved<T>, "urn" | "getProvider">;
```
Should this also be `Reso_lv_edResource<T extends...` instead of `Reso_vl_edResource<T ....`?

hausdorff commented:
Sweet jesus. Those spelling issues.

hausdorff commented:
@ellismg @lukehoban @justinvp one wrinkle about this polyfill + query stuff ... looking at the tests now, I _believe_ that with this change, everyone who uses Pulumi will have to have a line like this in their `tsconfig.json`:

```
{
    "compilerOptions": {
        "lib": ["dom", "es2015", "esnext.asynciterable"],
[...]
```

This does not seem ideal. What is our policy for this sort of thing? Is this even ok?

lukehoban commented:
> This does not seem ideal. What is our policy for this sort of thing? Is this even ok?

No - I do not think we want to require use of experimental TypeScript features.

hausdorff commented:
Ok. `@pulumi/query` fundamentally depends on those features. Should we be pulling all of this into a different package?

hausdorff commented:
One option is to move all the "query stuff" to `@pulumi/query` and simply expose a lighter version of `listResourceOutputs` in the core API.

lukehoban commented:
> @pulumi/query fundamentally depends on those features.

Do we have to - these are things that are not going to be familiar to any users.  Is there not value in a more traditional API design here?

> Should we be pulling all of this into a different package?

I don't think I follow what the options are.

> One option is to move all the "query stuff" to @pulumi/query and simply expose a lighter version of listResourceOutputs in the core API.

I thought that was already the plan of record.

hausdorff commented:
The plan of record was for `@pulumi/pulumi` to take a dependency on `@pulumi/query`. And we _did_ have a traditional API design but as you can see [earlier in this PR](https://github.com/pulumi/pulumi/pull/2630/files/2b94c9c49e10a14cd2ea61da98aa2b27e6089d3c#diff-4e9bafead188f5465daadeca3e442c59) the feedback was to use the iterator constructs _instead_ of this. So we went in that direction.

hausdorff commented:
Also, users don't have to know about async iterators. They are an implementation detail, mostly.

lukehoban commented:
> but as you can see [earlier in this PR](https://github.com/pulumi/pulumi/pull/2630/files/2b94c9c49e10a14cd2ea61da98aa2b27e6089d3c#diff-4e9bafead188f5465daadeca3e442c59) 

That link doesn't work for me - and I can't find where in the history here that was discussed.

hausdorff commented:
Sorry, here:https://github.com/pulumi/pulumi/pull/2630#pullrequestreview-228863725

justinvp commented:
> No - I do not think we want to require use of experimental TypeScript features.

To be clear, this is not an experimental TypeScript feature. Async Iteration was added in TypeScript 2.3 in April 2017 and it's part of the ES2018 standard. Instead of `esnext.asynciterable`, I believe `es2018.asynciterable` could be used to get the [type definitions](https://github.com/microsoft/TypeScript/blob/f944ed6d72577240b276e8bb03cc402033e1da6b/src/lib/es2018.asynciterable.d.ts).

> I _believe_ that with this change, everyone who uses Pulumi will have to have a line like this in their `tsconfig.json`:

If true, that is not ideal. All we're talking about is some strongly-typed definitions for the [`AsyncIterable` interface.](https://www.ecma-international.org/ecma-262/9.0/index.html#sec-asynciterable-interface). Can we just provide our own definitions for it?

```typescript
interface AsyncIterator<T> {
    next(value?: any): Promise<IteratorResult<T>>;
}

interface AsyncIterable<T> {
    [Symbol.asyncIterator](): AsyncIterator<T>;
}

interface AsyncIterableIterator<T> extends AsyncIterator<T> {
    [Symbol.asyncIterator](): AsyncIterableIterator<T>;
}
```

hausdorff commented:
Let's try that.

hausdorff commented:
@justinvp after fiddling around with this for a few hours, I am not sure it will be that easy. The problem is that `Symbol.asyncIterator` does not exist until ES2018. Also, it appears that we cannot supply our own definition of `Symbol` -- so something like `[MySymbol.asyncIterator](): AsyncIterableIterator<T>;` does not work. I'll keep poking at it but it's not super clear to me what the options are at this point.

hausdorff commented:
@justinvp @lukehoban @ellismg Ok, after much heartburn, I _think_ the best we can do is this: https://github.com/pulumi/pulumi-query/pull/7 Inside the commit message are links that explain the downsides to this approach. Re-declaring `SymbolConstructor` globally in pulumi-query does not seem amazing -- there may be a way to declare it scoped to where pulumi/pulumi uses it, but I am not knowledgable to make the call myself. If we do decide to go this route, this PR can be merged as soon as that PR merges. If not, we probably will want to not expose pulumi/query constructs as part of pulumi/pulumi. This probably means that users will have to do something like `query.from(pulumi.runtime.listResourceOutputs())` instead of simply `pulumi.runtime.listResourceOutputs()`. I don't like that option very much. What do you think?

justinvp commented:
> I _think_ the best we can do is this: [pulumi/pulumi-query#7](https://github.com/pulumi/pulumi-query/pull/7)

Thanks, Alex. That looks like the best approach, and is what other libraries are doing to address the same problem.

> Re-declaring SymbolConstructor globally in pulumi-query does not seem amazing

It's not amazing, but I actually don't think it's all that bad -- it's par for the course with TypeScript. Our declaration will merge fine with the declaration that users get when targeting ES6/ES2015 or later. We're just making the TypeScript compiler happy as we know the `Symbol.asynciterator` property will exist at runtime when using the library, either because the program is running on Node.js 10.x or higher (which already has it), or we've added the property on earlier versions of Node.js.

hausdorff commented:
Alright, thanks @justinvp. @ellismg if you're ok with this I think we should merge it and close out the related bugs.

hausdorff commented:
Final error looks like nothing to worry about. We're failing to resolve the stack name because the org name is missing, I'lll ask people about that tomorrow, should be quick to fix, and then we're ready to merge I think.
